content/blog/2020/07/2020-07-27-machine-learning-plugin-coding-phase2.adoc

```diff
@@ -0,0 +1,160 @@
+Machine learning plugin has successfully passed the end to end test. A feature for a code editor is needed for further discussion/analysis as we have done a simple editor that may be useful in other ways in the future. link:https://github.com/jenkinsci/machine-learning-plugin/pull/35[PR#35]
+
+==== Main features of Machine Learning plugin
+
+- Run Jupyter notebook, (Zeppelin) JSON and Python files
+- Run Python code directly
+- Convert Jupyter Notebooks to Python and JSON
+- Configure IPython kernel properties
+- Support to execute Notebooks/Python on Agent
+- Support for Windows and Linux
+
+====== Upcoming features
+
+- Extract graph/map/images from the code
+- Save artifacts according to the step name
+- Generate reports for corresponding build
+
+====== Future improvements
+
+- Usage of JupyterRestClient
+- Support for multiple language kernels
+
+==== Docker agent
+
+The following dockerfile can be used to build the docker container as an agent for machine learning plugin. This docker agent can be used to run  notebooks or python scripts.
+
+.Dockerfile
+
+```
+FROM jenkins/agent:latest
+
+MAINTAINER Loghi <loghijiaha@gmail.com>
+
+USER root
+
+RUN apt update && apt install --no-install-recommends python3 -y \
+    python3-pip \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt /requirements.txt
+
+RUN pip3 install --upgrade pip setuptools && \
+    pip3 install --no-cache-dir -r /requirements.txt && \
+    ln -sf /usr/bin/python3 /usr/bin/python && \
+    ln -sf /usr/bin/pip3 /usr/bin/pip
+
+USER jenkins
+```
+
+
+==== Ported to Python 3
+
+As discussed in the previous meeting, we concluded that the plugin should support Python 3 as Python 2.7+ has been deprecated since the beginning of 2020. Pull request for docker agent should be also ported to Python 3 support.
+
+==== Jupyter Rest Client API
+
+The Jupyter Notebook server API seemed to be promising that it can be also used to run notebooks and codes. There were 3 api implementations that were merged in the master. But we had to focus on what was proposed in the design document and had to finish all must-have issues/works. Jupyter REST client was left for future implementation. It is also a good start to contribute to the plugin from the community.
+
+
+==== Fixed bugs for running in agent
+
+There were a few bugs related to the file path of notebooks while building a job. The most problem was caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.
+
+==== R support as a future improvement
+
+This is what we tried to give a glimpse of knowledge that this plugin can be extended for multi language support in the future. There was a conclusion that the kernel should be selected dynamically using the mime type of notebook, instead of scripting the same code for each kernel.
```

Comment:
The mime type of a script I think (e.g. `train_model.r`, or `train_model.groovy`). We could identify the language, and use the right kernel to execute it (ideally leaving that up to Zeppelin and IPython as we discussed, instead of adding code for each language to the plugin).

For notebooks, we probably will need to parse/read the [magic commands](https://ipython.readthedocs.io/en/stable/interactive/magics.html), like `%%javascript`. And we will also need to consider the case when the notebook is completely written for JavaScript or for another language. In that case, the user would have chosen to use the JavaScript Kernel, and written in the cells without the magic line (that's how [I do to try JS features/libs](https://github.com/kinow/notebooks/blob/master/javascript/enumify/Enumify.ipynb)).

Reply:
Click on "Raw" on GitHub, and search for `kernelspec` and `language_info`, that's where that notebook is defining that it should run with JavaScript. That way I can have packages installed via `package.json`/npm.

Reply:
@kinow our plugin is capable of running magic commands as well. But we should consider running the train_model.r or train_model.groovy etc type of files. If we have time, we will plan for multi-language support in our plugin.